### DISCUSSION TITLE
USGS pilot FROST-Server implementation

### DISCUSSION THREAD

**Discussion by jkreft-usgs** [FraunhoferIOSB/FROST-Server#611] [Show and tell]
After many months resolving various scaling issues in our data transformation pipeline, the USGS has a pilot implementation of the Frost Server with live hydrologic data for the United States (mostly every 15 minutes) from over 14,000 Things and almost 56,000 datastreams. 

https://labs.waterdata.usgs.gov/sta/v1.1/

There is a lot more to come here, and more work to do to implement a more sophisticated data mapping.  At this point, I welcome any feedback on the service and would be happy to share additional implementation details.

**Comment by sgrellet** [comment]
Bonjour Jim,

That's great news !
Some questions from my side. Maybe those fall under the 'more sophisticated data mapping' you mention.

## Things
### 1°/ ultimateFeatureOfInterest
Are you planning to provide a link to the river / river segment monitored ? It's a use case our colleagues have brought back.
An example here : https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Datastreams(347)/Thing
Same here, we use FoI to provide the location (O&M V2 SF_SamplingPoint) where the measurement happens. We don't consider the monitored river / river segment as the FoI.

### 2°/ wells
It seems you also have Wells : https://labs.waterdata.usgs.gov/sta/v1.1/Things('USGS-373712086300203')
Is it a remnant of some tests or are you also pushing on this topic.

## Observation
Would be great to have quality information, we have be asked to add this.
https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Datastreams(347)/Observations

## ObservedProperties
I see entries such as "description": "NO3+NO2,water,insitu as N",  "Sample event", ...
Are you planning to also deploy ST API on Water Quality topics ? This would involve nice discussion on the FoI.

## Overall
As you know we have ST API (FROST) labs running on surface/ground water quantity and also surface water quality.
More than happy to join forces here :)

A bientôt

Sylvain
(Edited to fix formatting)

**Reply by jkreft-usgs** [reply]
## Things
### ultimateFeatureOfInterest
_Are you planning to provide a link to the river / river segment monitored ? It's a use case our colleagues have brought back.
An example here : https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Datastreams(347)/Thing
Same here, we use FoI to provide the location (O&M V2 SF_SamplingPoint) where the measurement happens. We don't consider the monitored river / river segment as the FoI._

```
{
  "description": "Le Rhin en Suisse [Rheinhalle]",
  "@iot.id": 174,
  "name": "Le Rhin en Suisse [Rheinhalle]",
  "properties": {
    "id": "https://iddata.eaufrance.fr/id/HydroStation/A021005050",
    "Commune": {
      "@type": "http://data.ign.fr/def/geofla#commune",
      "@id": "http://id.insee.fr/geo/commune/99140"
    },
    "relatedTo": [
      {
        "href": "http://iddata.eaufrance.fr/id/watershed/A021",
        "title": "Watershed"
      },
      {
        "href": "http://iddata.eaufrance.fr/id/watercourse/A---0000",
        "title": "Watercourse"
      }
    ]
  },
  "@iot.selfLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Things(174)",
  "MultiDatastreams@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Things(174)/MultiDatastreams",
  "HistoricalLocations@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Things(174)/HistoricalLocations",
  "Datastreams@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Things(174)/Datastreams",
  "TaskingCapabilities@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Things(174)/TaskingCapabilities",
  "Locations@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Things(174)/Locations"
}
```

I like this approach to linking to the hydrogeological context.  The Feature of Interest being tied to the observation itself doesn't work terribly well for semantic  relationships between the broader relationship of the *Thing* to the larger world, such as a river reach, lake,  watershed, or aquifer

Something I am curious about is how this paradigm would work with moving things.  Perhaps instead the above `relatedTo` block should instead be included as part of `Location` and `HistoricalLocation`



### wells
_It seems you also have Wells : https://labs.waterdata.usgs.gov/sta/v1.1/Things('USGS-373712086300203')
Is it a remnant of some tests or are you also pushing on this topic._

We are very much pushing on groundwater!  Hoping to include both instantaneous (15 minute data), daily stats, and most likely also discrete.  Eventually construction/lithology information as well, though not sure if that will be directly in STA or linked in an ancillary system.

## Observation
Would be great to have quality information, we have be asked to add this.
https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Datastreams(347)/Observations
https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Observations(173549184)
```
{
  "@iot.id": 173549184,
  "phenomenonTime": "2021-08-09T15:00:00.000Z",
  "result": 7340,
  "resultTime": "2021-08-09T15:00:00.000Z",
  "resultQuality": [
    {
      "nameOfMeasure": "DQ_Status",
      "DQ_Result": {
        "code": "http://id.eaufrance.fr/nsa/510#16",
        "label": "Validated"
      }
    },
    {
      "nameOfMeasure": "DQ_Qualification",
      "DQ_Result": {
        "code": "http://id.eaufrance.fr/nsa/515#20",
        "label": "Good"
      }
    }
  ],
  "@iot.selfLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Observations(173549184)",
  "Datastream@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Observations(173549184)/Datastream",
  "MultiDatastream@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Observations(173549184)/MultiDatastream",
  "FeatureOfInterest@iot.navigationLink": "https://sta4hydrometry.brgm-rec.fr/FROST-Server/v1.1/Observations(173549184)/FeatureOfInterest"
}
```
Yes! At this point, we are going to be putting out status code (such as Approved, Provisional, etc) as well as masks such as ("Ice affected" and "equipment failure").  We are not planning (at this point) to put out qualifers  (e.g. good) at this point- not really part of our data model most of the time.
 
## ObservedProperties
_I see entries such as "description": "NO3+NO2,water,insitu as N",  "Sample event", ...
Are you planning to also deploy ST API on Water Quality topics ? This would involve nice discussion on the FoI._

We are!  We are doing a LOT of work in discrete water quality, and also collect many different continuous water quality samples, which we are also reconsidering right now.  Exactly how to effectively link discrete and continuous water quality is an area that we are arvily working towards.

## Overall
_As you know we have ST API (FROST) labs running on surface/ground water quantity and also surface water quality.
More than happy to join forces here :)_

Same here!

**Reply by sgrellet** [reply]
A bit late back on that thread. 
We are on the same page as often. And I feel ST API implementations in this domain now reach a critical mass that's worth sharing with the community.
Let's have a ST API session in the OGC HydroDWG meetings about all this.

**Comment by hylkevds** [comment]
Cool, thank you for sharing!
As a quick test I've added your data source to our demo WebGenesis front end: https://api4inspire.k8s.ilt-dmz.iosb.fraunhofer.de/servlet/is/199/
Loading takes a while, since it loads all Things at the start. Just click on the layer overview to see the loading progress.

**Reply by jkreft-usgs** [reply]
Cool! Thank you for adding it in. I have shared this out internally.
